enhancements/cvo/upgrade_preflight_checks.md

```diff
@@ -0,0 +1,118 @@
+* Administrators will be able to run preflight checks without running the upgrade.
+* Give Operators (CVO managed) ability to add checks to check if they can safely upgrade to a given release.
+* Preflight checks would be idempotent in nature.
+
+### Non-Goals
+
+* This enhancement does not commit particular operators for providing preflight implementations or discuss the operator-specific logic that would be executed by preflight checks.
+* The preflight checks should not be used as a migration mechanism.
+
+## Proposal
+
+Operators need to provide a different bin path for the preflight checks i.e. /bin/preflight.
+The binpath need to exist inside of the operator image.
+Cluster version operator (CVO) will be running a single job per operator for the preflight checks.
+The binpath /bin/preflight must be executable and it can be symbolic link to the operator binary or a script or It can be non existent.
+CVO will not fail if the path does not exist in an operator image but it would assume that preflight checks are not implemented for the respective operator.
+Also there would be a possible way to skip the preflight checks if someone wants to skip it explicitly.
+
+### User Stories
+
+#### Story 1
+
+Some customers would prefer to disable automatic credential provisioning by Cloud credential operator (CCO) and instead manage it themselves.
+However situations can arise during an upgrade where credentials have changed and the upgrade will be block and go into a failed state, with little visibility as to why or how it can be fixed.
+With the help of preflight checks CCO can check if they can safely upgrade to a given release before doing the actual upgrade.
+Communicate that an upgrade is not possible as permissions have changed and point administrator with right pointers to fix the issue.
+This would allow users to avoid broken upgrades when credentials have changed and provides them right information to move forward.
+
+### Implementation Details
+
+CVO will run the preflight check from the new operator image to which the current operator image is trying to upgrade.
+Because the current version of operator does not have knowledge of what new code changes the new version bringing with it.
+So its the new version's responsibility to figure out if the current image can update safely to the new image.
+However CVO need to go through signature verification to pass on the target release before we launch preflights
+
+The preflight jobs would run in the same namespace as operators as it would need same RBAC as the operators.
+
+#### Triggering execution of preflight checks
+
+We should be able to run the preflight check as below
+
+* Preflight checks would run right after the update is selected, as we currently do for preconditions.
+* Run these on-demand for cluster admins. As it would help for them in the process of planning on updating to a specific version.
+
+#### Reporting results from preflight checks
```

Comment:
AFAIU preflight checks would run in the current to be updated environment. What about the case where the operator needs the new environment to run successfully, what information can we get about the update? 

Especially in the case of out-of-tree drivers, the hardware enablement operators would need to know e.g. the new kernel version or operating system version. Would this be possible to extract in a preflight check?

Reply:
This pre-flight mechanism is for CVO-managed operators only. In your example, the hardware enablement operators would know during CI whether or not they are compatible with the given OS version. Third party operators could probably make use of some future mechanism, though I would probably lean toward MCD/MCO for your example since it is so closely tied to the OS.

Reply:
@crawford We have this enhancement for out-of-tree driver support in OpenShift, which would make SRO a CVO-managed operator: https://github.com/openshift/enhancements/pull/357/. This is the case I am "worried" about.  

I am totally on your side that a CI should catch that, be it our or third party enablement operator. Since we're trying to create a general solution here we want also to catch cases where the drivers are not part of a CI/CD pipeline. 

My thinking was to create some preflight checks so that customers do not upgrade if the drivers are not compatible and render their cluster unusable. 

In the case of hardware accelerators, I would like to know the kernel and os version and if the kernel-devel package is available. In the past, I have seen cases where e.g. we had a bugfix kernel without the kernel-devel and kernel-headers packages on an entitled cluster, where those were part of the machine-os-content and not the official rhel repositories. 

/cc @mrunalp